File: Backtracking/generate_parentheses.cpp
// Generate all combinations of valid parentheses
#include <iostream>

using namespace std;

void solve(string res, int open, int close, int size) {
  if(res.size() == size * 2) {
    cout << res << endl;
    return;
  }

  if(open < size)
    solve(res + "(", open + 1, close, size);
  if(close < open)
    solve(res + ")", open, close + 1, size);
}

int main() {

  int n = 4;

  solve("", 0, 0, n);


  return 0;
}

File: Backtracking/generate_parentheses.go
// https://leetcode.com/problems/generate-parentheses/description

func solve(n, opened, closed int, curr string, ans *[]string) {
    if opened == n && closed == n {
        *ans = append(*ans, strings.Clone(curr))
        return
    }

    if opened == n {
        solve(n, opened, closed+1, curr + ")", ans)
        return
    }

    if opened > closed {
        solve(n, opened+1, closed, curr + "(", ans)
        solve(n, opened, closed+1, curr + ")", ans)
    }

    if opened == closed {
        solve(n, opened+1, closed, curr + "(", ans)
    }
}

func generateParenthesis(n int) []string {
    ans := []string{}

    solve(n, 0, 0, "", &ans)

    return ans
}
